src/gluonts/dataset/loader.py

```diff
@@ -58,52 +66,53 @@ def __init__(
         batch_size: int,
         ctx: mx.Context,
         dtype: DType = np.float32,
+        num_workers: int = None,
+        cyclic: bool = False,
+        **kwargs
     ) -> None:
+        # conversion from iterator to list
         self.batch_size = batch_size
         self.ctx = ctx
         self.dtype = dtype
         self.is_train = is_train
-        self.dataset = dataset
         self.transform = transform
-
-    @property
-    def stream(self) -> Iterable:
-        return self.transform(self.dataset, is_train=self.is_train)
-
-    def make_batch_iter(self):
-        batches = batcher(self.stream, self.batch_size)
-        stack = functools.partial(dct_reduce, self.stack)
-        return map(stack, batches)
-
-    def stack(self, xs):
-        if isinstance(xs[0], np.ndarray):
-            data = np.asarray(xs)
-            if data.dtype.kind == "f":
-                data = data.astype(self.dtype)
-            return mx.nd.array(data, dtype=data.dtype, ctx=self.ctx)
-
-        if isinstance(xs[0], mx.nd.NDArray):
-            return mx.nd.stack(*xs)
-
-        if isinstance(xs[0], list):
-            return list(self.stack(t) for t in zip(*xs))
-
-        if isinstance(xs[0], tuple):
-            return tuple(self.stack(t) for t in zip(*xs))
-
-        return xs
+        self.cyclic = cyclic
+
+        # TODO: think about what a good value is, probably 0, and if multiprocessing=True, then what is below
+        if num_workers is None:
+            self.num_workers = min(len(list(dataset)), int(cpu_count() / 2))
+        else:
+            assert num_workers <= len(
+                list(dataset)
+            ), "Cannot have more workers than dataset entries currently."
+            self.num_workers = num_workers
+
+        self.num_workers = 0
```

Comment:
This is probably left over from testing and should be removed?

Reply:
yes

Reply:
I alway ran pytest with what I have above, and with 0 to cover both cases

Reply:
> Thanks @AaronSpieler, this is great.
> 
> I think we could write one small test to check that the `TrainDataLoader` gives (up to permutation) the same batches for `num_workers > 1` and `num_workers = 0` when running some simple deterministic transformation and then merge this change.

I can tell you already that we will get probably slightly different results: The instance splitter probabilistically samples the number of split points in a single time series, if we have split the dataset by time series, and run a transformation over each subset, the sampling will most probably give different results:
The only thing that can be realistically tested is: is evert time series transformed only once?

Reply:
We can use the `UniformSplitSampler` and set the probability to `1.0` for the test. Then every possible split is selected.

Reply:
I see, that should of course yield the exact same result.